Add helm option enableEgressGateway

anfernee opened the pull request:
The option turns on the egress gateway support, which allows users to
specify Egress NAT policies to redirect and SNAT the egress traffic that
leaves the cluster. It's false by default.

Towards: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>

kaworu reviewed (changes requested):
Hello @anfernee and thanks for the PR!

Question: should the flag stay hidden if we're exposing it through Helm?

[This failure](https://github.com/cilium/cilium/pull/15777/checks?check_run_id=2384965259) seems legit to me, and the [helm-docs](https://github.com/cilium/cilium/pull/15777/checks?check_run_id=2384965652) reordered the `enableEgressGateway` in the README.md for some reason (`make -C install/kubernetes` should fix it).


/cc @joestringer as it seems you have context about https://github.com/cilium/cilium/pull/14723.

aanm commented on install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml:
```diff
@@ -104,6 +104,7 @@ rules:
   - ciliumlocalredirectpolicies
   - ciliumlocalredirectpolicies/status
   - ciliumlocalredirectpolicies/finalizers
+  - ciliumegressnatpolicies
```
We need `ciliumlocalredirectpolicies/status`. I'm not so sure about `ciliumlocalredirectpolicies/finalizers`. But we should probably put it to be on the safe side?

anfernee replied:
ciliumegressnatpolicies doesn't have `status` field. but I guess there is no harm adding it though. will follow the pattern here.

aanm replied:
@anfernee if that is the case then we should remove it from the CRD as it is currently defined in there https://github.com/cilium/cilium/blob/a783f7c55d34b647e751afe6fc7e892b1a809485/examples/crds/ciliumegressnatpolicies.yaml#L184-L185
https://github.com/cilium/cilium/blob/a783f7c55d34b647e751afe6fc7e892b1a809485/pkg/k8s/apis/cilium.io/v2alpha1/types.go#L27

pchaigno commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -677,6 +677,10 @@ data:
   disable-cnp-status-updates: "true"
 {{- end }}
 
+{{- if .Values.enableEgressGateway }}
```
That's inconsistent with other names. We should probably go for `egressGateway.enabled=true` instead. Unless we're sure we won't ever need other subflags to `egressGateway` in which case just `egressGateway=true` is enough.

pchaigno reviewed (changes requested):
Oops, almost missed something :-)

anfernee committed:
api: Remove status from CRD ciliumegressnatpolicies

The new type doesn't have status field, so remove it.

Towards: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>

anfernee committed:
daemon: Unhide --enable-egress-gateway

Towards: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>

anfernee committed:
Add helm option enableEgressGateway

The option turns on the egress gateway support, which allows users to
specify Egress NAT policies to redirect and SNAT the egress traffic that
leaves the cluster. It's false by default.

Towards: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>
